File: dart/maximum-average-pass-ratio.dart
import "package:collection/collection.dart";

class Solution {
  double maxAverageRatio(List<List<int>> classes, int extraStudents) {
    PriorityQueue<(int, int)> queue = HeapPriorityQueue<(int, int)>(
      ((int, int) class1, (int, int) class2) =>
        brilliantStudentImpact(class2).compareTo(brilliantStudentImpact(class1))
    );
    for (final clazz in classes) {
      queue.add((clazz[0], clazz[1]));
    }
    for (int i = 0; i < extraStudents; i++) {
      var (pass, total) = queue.removeFirst();
      queue.add((pass + 1, total + 1));
    }
    double sum = 0;
    while (queue.isNotEmpty) {
      var (pass, total) = queue.removeFirst();
      sum += pass.toDouble() / total;
    }
    return sum / classes.length;
  }

  double brilliantStudentImpact((int, int) clazz) {
    var (pass, total) = clazz;
    return (pass + 1.0) / (total + 1) - pass.toDouble() / total;
  }
}

File: python3/maximum-average-pass-ratio.py
from typing import List

import heapq


class Solution:
    """
    >>> round(Solution().maxAverageRatio([[1,2],[3,5],[2,2]], 2), 5)
    0.78333
    >>> round(Solution().maxAverageRatio([[2,4],[3,9],[4,5],[2,10]], 4), 5)
    0.53485
    """

    def maxAverageRatio(self, classes: List[List[int]], extraStudents: int) -> float:
        heap = []
        for passed, total in classes:
            heapq.heappush(
                heap,
                (passed / total - (passed + 1) / (total + 1), passed, total)
            )
        for _ in range(extraStudents):
            _improve_ratio, passed, total = heapq.heappop(heap)
            heapq.heappush(
                heap,
                ((passed + 1) / (total + 1) - (passed + 2) / (total + 2), passed + 1, total + 1)
            )
        pass_ratio = 0.0
        for _, passed, total in heap:
            pass_ratio += passed / total
        return pass_ratio / len(heap)
